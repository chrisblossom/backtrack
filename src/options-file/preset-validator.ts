import Joi from '@hapi/joi';
import { BacktrackConfig } from '../types';
import { packageJsonSchema } from './package-json-schema';

const copyFilesSchema = Joi.object({
	src: Joi.string().required(),
	dest: Joi.string().required(),
	allowChanges: Joi.boolean(),
	ignoreUpdates: Joi.boolean(),
});

const filesOptionsSchema = Joi.object({
	skip: Joi.array().items(Joi.string()).single(true),

	allowChanges: [
		//
		Joi.array().items(Joi.string()).single(true),
		Joi.boolean(),
	],
	ignoreUpdates: [
		Joi.array().items(Joi.string()).single(true),
		Joi.boolean(),
	],

	makeDirs: Joi.array().items(Joi.string()).single(true),
});

const filesSchema = Joi.array()
	.items(copyFilesSchema, filesOptionsSchema)
	.single(true)
	.unique()
	.options({
		allowUnknown: false,
	})
	.label('files');

const cmdType = Joi.string().label('cmd');
const fnType = Joi.func().label('function');
const objType = Joi.object({
	name: Joi.string().required(),
	task: [
		fnType.required(),
		cmdType.required(),
	],
});

const functionLifecycles = Joi.array()
	.items([
		cmdType,
		fnType,
		objType,
		Joi.array().items(cmdType, fnType, objType),
	])
	.single(true);

const cleanCopySchema = Joi.array()
	.items(
		Joi.object({
			dest: Joi.string().required(),
			src: Joi.string().required(),
			hash: Joi.boolean(),
		}),
	)
	.single(true)
	.min(1);

const cleanSchema = Joi.array()
	.items(
		Joi.object({
			del: Joi.array().items(Joi.string()).single(true).min(1),
			makeDirs: Joi.array().items(Joi.string()).single(true).min(1),
			copy: cleanCopySchema,
		}).min(1),
	)
	.single(true)
	.unique()
	.options({
		allowUnknown: false,
	})
	.label('clean');

const configSchema = Joi.array()
	.items(
		Joi.object({}).options({
			allowUnknown: true,
		}),
	)
	.single(true)
	.label('config');

const resolveSchema = Joi.object().pattern(/.*/, Joi.string()).label('resolve');

function generateSchema(lifecycles: BacktrackConfig): Joi.ObjectSchema {
	const managedLifecycles = {
		clean: cleanSchema,
		files: filesSchema,
		packageJson: packageJsonSchema,
		config: configSchema,
		resolve: resolveSchema,
	};

	/**
	 * Although all lifecycles can be false to be disabled, they are removed before validation
	 * Do not not test for this because it will return poor error messages
	 */
	type ManagedLifecycles = keyof typeof managedLifecycles;
	const allLifeCyclesKeys = Object.keys(lifecycles) as ManagedLifecycles[];
	const allLifecycles = allLifeCyclesKeys.reduce(
		(acc, lifecycle: ManagedLifecycles) => {
			const rule = managedLifecycles[lifecycle]
				? managedLifecycles[lifecycle]
				: functionLifecycles.label(lifecycle);

			return {
				...acc,
				[lifecycle]: [rule],
			};
		},
		{},
	) as Record<ManagedLifecycles, Joi.Schema>;

	const result = Joi.object({
		...allLifecycles,
		presets: Joi.array()
			.items(
				Joi.string(),
				Joi.array().ordered(Joi.string().required(), Joi.object()),
			)

			.single(true)
			.unique()
			.label('presets'),
	}).required();

	return result;
}

interface Args {
	value: BacktrackConfig;
}

function presetValidator({ value = {} }: Args = { value: {} }): void {
	const schema = generateSchema(value);
	const isValid = Joi.validate(value, schema);

	if (isValid.error) {
		throw new Error(isValid.error.annotate());
	}
}

export { presetValidator };
