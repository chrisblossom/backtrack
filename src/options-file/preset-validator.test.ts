import { presetValidator } from './preset-validator';

describe('presetValidator', () => {
	test('handles undefined', () => {
		const validated = presetValidator();

		expect(validated).toEqual(undefined);
	});

	test('preset passes', () => {
		const value = require('./__sandbox__/preset-01');

		const validated = presetValidator({ value });

		expect(validated).toEqual(undefined);
	});

	test('preset passes with options', () => {
		const value = {
			presets: [
				[
					'./__sandbox__/preset-01',
					{ options: true },
				],
			],
		};

		// @ts-ignore
		const validated = presetValidator({ value });

		expect(validated).toEqual(undefined);
	});

	test('preset validates custom lifecycles', () => {
		const value = {
			invalid1: new Date('2017-12-05T18:02:11.869Z'),
		};

		try {
			expect.hasAssertions();
			// @ts-ignore
			presetValidator({ value });
		} catch (error) {
			expect(error).toMatchSnapshot();
		}
	});

	test('preset allows no valid options', () => {
		const value = {};

		const validated = presetValidator({ value });
		expect(validated).toEqual(undefined);
	});

	test('allows empty presets', () => {
		const value = {
			dev: [],
			clean: [],
			build: [],
			lint: [],
			test: [],
			files: [],
			format: [],
			packageJson: [],
			config: [],
		};

		const validated = presetValidator({ value });
		expect(validated).toEqual(undefined);
	});

	test('fails with unknown key', () => {
		const value = {
			clean: [
				{
					del: ['*'],
					makeDirs: ['dir'],
					excess: 'something',
				},
			],
		};

		try {
			expect.hasAssertions();
			// @ts-ignore
			presetValidator({ value });
		} catch (error) {
			expect(error).toMatchSnapshot();
		}
	});

	test('fails no options - makeDirs', () => {
		const value = {
			clean: [
				{
					makeDirs: [''],
				},
			],
		};

		try {
			expect.hasAssertions();
			// @ts-ignore
			presetValidator({ value });
		} catch (error) {
			expect(error).toMatchSnapshot();
		}
	});

	test('fails no options - del', () => {
		const value = {
			clean: [
				{
					del: [''],
				},
			],
		};

		try {
			expect.hasAssertions();
			// @ts-ignore
			presetValidator({ value });
		} catch (error) {
			expect(error).toMatchSnapshot();
		}
	});

	test('success on clean', () => {
		const value = {
			clean: [
				{
					copy: {
						src: 'static',
						dest: 'static',
					},
				},
				{
					del: ['**'],
					makeDirs: ['fake/dir'],
				},
			],
		};

		// @ts-ignore
		const validated = presetValidator({ value });

		expect(validated).toEqual(undefined);
	});

	test('success with files options', () => {
		const value = {
			files: [
				{
					src: 'file1.js',
					dest: 'file1.js',
				},
				{
					src: 'nested/inside.js',
					dest: 'nested/inside.js',
					allowChanges: true,
				},
				{
					src: 'nested/inside.js',
					dest: 'nested/inside.js',
					ignoreUpdates: true,
				},
				{
					src: 'nested/inside.js',
					dest: 'nested/inside.js',
					allowChanges: true,
					ignoreUpdates: true,
				},
				{
					src: 'nested/other.js',
					dest: 'nested/other.js',
				},
				{
					skip: ['file1.js'],
					allowChanges: ['file1.js'],
					makeDirs: ['dist'],
					ignoreUpdates: true,
				},
				{
					skip: ['file1.js'],
				},
				{
					makeDirs: ['dist'],
				},
				{
					allowChanges: true,
				},
				{
					ignoreUpdates: true,
				},
			],
		};

		// @ts-ignore
		const validated = presetValidator({ value });
		expect(validated).toEqual(undefined);
	});

	test('success with resolve', () => {
		const value = {
			resolve: {
				'some-npm-pkg': '/path/to/pkg',
			},
		};

		// @ts-ignore
		const validated = presetValidator({ value });
		expect(validated).toEqual(undefined);
	});
});
