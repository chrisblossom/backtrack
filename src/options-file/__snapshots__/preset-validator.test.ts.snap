// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presetValidator fails no options - del 1`] = `
[Error: {
  "clean": [
    {
      "del": [
        "" [31m[1][0m
      ]
    }
  ]
}
[31m
[1] "0" is not allowed to be empty[0m]
`;

exports[`presetValidator fails no options - makeDirs 1`] = `
[Error: {
  "clean": [
    {
      "makeDirs": [
        "" [31m[1][0m
      ]
    }
  ]
}
[31m
[1] "0" is not allowed to be empty[0m]
`;

exports[`presetValidator fails with unknown key 1`] = `
[Error: {
  "clean": [
    {
      "del": [
        "*"
      ],
      "makeDirs": [
        "dir"
      ],
      "excess" [31m[1][0m: "something"
    }
  ]
}
[31m
[1] "excess" is not allowed[0m]
`;

exports[`presetValidator preset validates custom lifecycles 1`] = `
[Error: {
  "invalid1" [31m[1][0m: "2017-12-05T18:02:11.869Z"
}
[31m
[1] single value of "invalid1" does not match any of the allowed types[0m]
`;
